topdown/tokens.go

```diff
@@ -864,8 +864,90 @@ func builtinJWTDecodeVerify(a ast.Value, b ast.Value) (v ast.Value, err error) {
 	return arr, nil
 }
 
+func extractUnverifiedPayloadAsAST(token ast.Value) (ast.Value, error) {
+	jwtToken, err := decodeJWT(token)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := builtinBase64UrlDecode(ast.String(jwtToken.payload))
+	if err != nil {
+		return nil, fmt.Errorf("JWT payload had invalid encoding: %v", err)
+	}
+
+	payload, err := extractJSONObject(string(p.(ast.String)));
+	if err != nil {
+		return nil, err
+	}
+
+	return payload, nil
+}
+
+// Implements full JWT decoding, validation and verification.
+func builtinOpenIdConnectTokenVerifyAndParse(a ast.Value, b ast.Value) (v ast.Value, err error) {
+	// io.openid.verify(string, [trusted_idp's])
+	//
+	// If valid is true iff token is valid.
+	//
+	// Decoding errors etc are returned as errors.
+	ret := make(ast.Array, 3)
+	ret[0] = ast.BooleanTerm(false)    // By default, not verified
+	ret[1] = ast.NewTerm(ast.NewObject()) // The parsed payload
+
+	// Grab the token as a string.
+	var token *string
+	if token, err = getString(a); err != nil {
+		return
+	}
+
+	// Parse the trusted issuers.
+	var trustedIssuers []*string
```

Comment:
You are correct 100%. They could be normal `string` types. It's safer to use `string` directly, to eliminate the possibility of `nil` dereference. Though I was thinking it may be faster to use pointers (to reduce the number of string copies).